mpt/impl/smtp/cassandra-rabbitmq-object-storage/src/test/java/org/apache/james/mpt/smtp/CassandraRabbitMQSwiftSmtpTestRule.java

```diff
@@ -18,120 +18,33 @@
  ****************************************************************/
 package org.apache.james.mpt.smtp;
 
-import java.util.Iterator;
-
 import org.apache.commons.configuration.DefaultConfigurationBuilder;
 import org.apache.james.CassandraJamesServerMain;
 import org.apache.james.CleanupTasksPerformer;
 import org.apache.james.GuiceJamesServer;
 import org.apache.james.backend.rabbitmq.DockerRabbitMQSingleton;
 import org.apache.james.backends.cassandra.init.configuration.ClusterConfiguration;
 import org.apache.james.dnsservice.api.DNSService;
-import org.apache.james.dnsservice.api.InMemoryDNSService;
 import org.apache.james.modules.TestRabbitMQModule;
 import org.apache.james.modules.TestSwiftBlobStoreModule;
 import org.apache.james.modules.blobstore.BlobStoreChoosingModule;
-import org.apache.james.modules.protocols.ProtocolHandlerModule;
-import org.apache.james.modules.protocols.SMTPServerModule;
-import org.apache.james.modules.protocols.SmtpGuiceProbe;
 import org.apache.james.modules.protocols.SmtpGuiceProbe.SmtpServerConnectedType;
 import org.apache.james.modules.rabbitmq.RabbitMQModule;
 import org.apache.james.modules.server.CamelMailetContainerModule;
-import org.apache.james.mpt.api.Continuation;
-import org.apache.james.mpt.api.Session;
-import org.apache.james.mpt.monitor.SystemLoggingMonitor;
-import org.apache.james.mpt.session.ExternalSessionFactory;
 import org.apache.james.queue.api.MailQueueItemDecoratorFactory;
 import org.apache.james.queue.api.RawMailQueueItemDecoratorFactory;
 import org.apache.james.server.core.configuration.Configuration;
 import org.apache.james.util.Host;
-import org.apache.james.util.Port;
-import org.apache.james.utils.DataProbeImpl;
-import org.junit.rules.TemporaryFolder;
-import org.junit.rules.TestRule;
-import org.junit.runner.Description;
-import org.junit.runners.model.Statement;
 
-import com.google.common.base.Preconditions;
-import com.google.common.base.Splitter;
-import com.google.inject.Module;
 import com.google.inject.util.Modules;
 
-public class CassandraRabbitMQSwiftSmtpTestRule implements TestRule, SmtpHostSystem {
-
-    private static final Module SMTP_PROTOCOL_MODULE = Modules.combine(
-        new ProtocolHandlerModule(),
-        new SMTPServerModule());
-
-    private final Host cassandraHost;
-    private final SmtpServerConnectedType smtpServerConnectedType;
-
-    private TemporaryFolder folder;
-    private GuiceJamesServer jamesServer;
-    private InMemoryDNSService inMemoryDNSService;
-    private ExternalSessionFactory sessionFactory;
+public class CassandraRabbitMQSwiftSmtpTestRule extends SmtpTestRule {
```

Comment:
I'd like to avoid inheritance when possible. Here, you just need to provide a function to SmtpTestRule that creates the server when given the right information (maybe you need a bit more but I'd like to keep using builders instead of non-composable hierarchies)

Reply:
I agree, but here `createJamesServer` needs a lot of attributes to work, IMO it does not worth the effort.

Reply:
Better design, thanks